### ISSUE TITLE
test_conv2d sometimes fails.

### ISSUE THREAD

**Issue by tommydcjung** [bespoke-silicon-group/bsg_replicant#496]
```
INFO:    (M, N, H, W, P, Sx, Sy, Bx, By) = (22, 32, 1, 1, 4, 6, 8, 4, 7)
ERROR:   Mismatch: B[0, 0] = 0.000000000 = 0x800000 	 Expected: 0.000000000 = 0x0
ERROR:   Mismatch: B[0, 1] = 0.000000000 = 0x800000 	 Expected: 0.000000000 = 0x0
ERROR:   Mismatch: B[0, 2] = 0.000000000 = 0x800000 	 Expected: 0.000000000 = 0x0
ERROR:   Mismatch: B[0, 3] = 0.000000000 = 0x800000 	 Expected: 0.000000000 = 0x0
ERROR:   Mismatch: B[1, 0] = 0.000000000 = 0x800000 	 Expected: 0.000000000 = 0x0
ERROR:   Mismatch: B[1, 1] = 0.000000000 = 0x800000 	 Expected: 0.000000000 = 0x0
ERROR:   Mismatch: B[1, 2] = 0.000000000 = 0x800000 	 Expected: 0.000000000 = 0x0
ERROR:   Mismatch: B[1, 3] = 0.000000000 = 0x800000 	 Expected: 0.000000000 = 0x0
ERROR:   Mismatch: B[3, 1] = 0.000000000 = 0x800000 	 Expected: 0.000000000 = 0x0
ERROR:   Mismatch: B[3, 2] = 0.000000000 = 0x800000 	 Expected: 0.000000000 = 0x0
ERROR:   Mismatch: B[5, 0] = 0.000000000 = 0x800000 	 Expected: 0.000000000 = 0x0
ERROR:   Mismatch: B[5, 1] = 0.000000000 = 0x800000 	 Expected: 0.000000000 = 0x0
ERROR:   Mismatch: B[6, 3] = 0.000000000 = 0x800000 	 Expected: 0.000000000 = 0x0
BSG ERR: Vectors don't match!
BSG REGRESSION TEST FAILED

```

**Comment by drichmond** [comment]
How do you reproduce this? Does it happen on the first try?

**Comment by tommydcjung** [comment]
probably on a second try. Seems weird.

**Comment by tommydcjung** [comment]
also the result varies each time.
```
INFO:    (M, N, H, W, P, Sx, Sy, Bx, By) = (8, 11, 8, 4, 5, 8, 8, 2, 3)
ERROR:   Mismatch: B[0, 0] = 0.000000000 = 0x800000 	 Expected: 0.000000000 = 0x0
ERROR:   Mismatch: B[1, 0] = 0.000000000 = 0x800000 	 Expected: 0.000000000 = 0x0
ERROR:   Mismatch: B[1, 1] = 0.000000000 = 0x800000 	 Expected: 0.000000000 = 0x0
ERROR:   Mismatch: B[2, 1] = 0.000000000 = 0x800000 	 Expected: 0.000000000 = 0x0
BSG ERR: Vectors don't match!
BSG REGRESSION TEST FAILED

```

**Comment by drichmond** [comment]
Cool, thanks @tommydcjung!

**Comment by mrutt92** [comment]
I have seen this too. @tommydcjung have what's the machine configuration you have observed this on.

**Comment by tommydcjung** [comment]
Running on default mode in master

**Comment by save-buffer** [comment]
I think this is a problem with the float comparison code. It's comparing 1.17549435E-38 against 0. Not super sure how to fix this - everywhere I look says that there's no silver bullet to handle all cases. I think the current comparison code is fine with smaller, more normal-looking floats. I've used this as a resource: 
https://floating-point-gui.de/errors/comparison/

I can get it to consistently pass by generating smaller values, on the order of 1e4.

**Comment by dpetrisko** [comment]
As a quick fix, you could check if operands are subnormal and pass the comparison with a warning?

**Comment by drichmond** [comment]
Why not use sum-squared-error? It's not perfect, but as long as your error is subnormal you should be fine? Or below some epsilon

This is how I do it in my matrix multiply code

**Comment by save-buffer** [comment]
SSE is also a good idea, I'll switch to that for now.

**Comment by drichmond** [comment]
Addressed in #499
